spec/usecases/reactions/update_materials_spec.rb

```diff
@@ -0,0 +1,106 @@
+  let(:molfile) { File.read(Rails.root + 'spec/fixtures/test_2.mol') }
+  let(:starting_materials) do
+    {
+      'starting_materials' => [
+        {
+          # hits existing_sample branch
+          'id' => sample.id,
+          'name' => 'starting_material',
+          'target_amount_unit' => 'mg',
+          'target_amount_value' => 75.09596,
+          'equivalent' => 1,
+          'reference' => false,
+          'is_new' => false,
+          'molfile' => molfile,
+          'container' => root_container
+        }
+      ]
+    }
+  end
+  let(:reactants) do
+    {
+      'reactants' => [
+        # hits subsample branch
+        'target_amount_unit' => 'mg',
+        'target_amount_value' => 86.09596,
+        'equivalent' => 2,
+        'reference' => false,
+        'is_new' => true,
+        'molfile' => molfile,
+        'container' => root_container,
+        'parent_id' => sample.id # gets named after parent, hence no name specified
+      ]
+    }
+  end
+  let(:products) do
+    {
+      'products' => [
+        {
+          # hits new_sample branch
+          'name' => 'product',
+          'target_amount_unit' => 'mg',
+          'target_amount_value' => 99.08304,
+          'equivalent' => 5.5,
+          'reference' => false,
+          'is_new' => true,
+          'molfile' => molfile,
+          'container' => root_container
+        }
+      ]
+    }
+  end
+  let(:materials) do
+    {
+      'solvents' => [
+        # hits new_sample branch
+        'name' => 'solvent',
+        'target_amount_unit' => 'mg',
+        'target_amount_value' => 76.09596,
+        'equivalent' => 1,
+        'reference' => true,
+        'is_new' => true,
+        'molfile' => molfile,
+        'container' => root_container
+      ]
+    }.merge(starting_materials, reactants, products)
+  end
+
+  it 'updates existing sample' do
+    described_class.new(reaction, starting_materials, user).execute!
+    expect(Sample.all.size).to eq(1) # RSpec creates `sample` fixture, and ReactionHelpers update it
+    expect(Sample.find_by(name: 'starting_material').target_amount_value).to eq(75.09596)
+  end
+
+  it 'creates sample for new materials with parent ID that are not products' do
+    described_class.new(reaction, reactants, user).execute!
+    expect(Sample.all.size).to eq(2) # RSpec creates `sample` fixture (parent), ReactionHelpers derive new sample from it
+    expect(Sample.find_by(short_label: 'reactant').target_amount_value).to eq(86.09596)
+  end
+
+  it 'creates sample for new materials that are products' do
+    described_class.new(reaction, products, user).execute!
+    expect(Sample.all.size).to eq(1) # ReactionHelpers create new sample
+    expect(Sample.find_by(name: 'product').target_amount_value).to eq(99.08304)
+  end
+
+  it 'associates reaction with materials' do
+    described_class.new(reaction, materials, user).execute!
+    expect(ReactionsSample.all.size).to eq(4)
```

Comment:
You can use `ReactionsSample.count` to just to a count query without instantiating all fetched objects